packages/cli/src/services/package-manager/common.ts

```diff
@@ -15,18 +16,52 @@ const ensureProjectDir = (processService: ProcessService, projectDirRef: Ref.Ref
     return projectDir
   })
 
+/**
+ * Checks that the script exists in the package.json file
+ */
+const checkScriptExists = (processService: ProcessService, fileSystemService: FileSystemService, scriptName: string) =>
+  gen(function* ($) {
+    const { cwd } = processService
+    const { readFileContents } = fileSystemService
+    const pwd = yield* $(cwd())
+    const packageJson = yield* $(readFileContents(`${pwd}/package.json`))
+    const packageJsonContents = JSON.parse(packageJson)
+    if (!packageJsonContents.scripts || !packageJsonContents.scripts[scriptName]) {
+      return false
+    }
+    return true
+  })
+
+const effectfulRun = (processService: ProcessService, command: string, projectDirRef: Ref.Ref<string>) => {
+  const { exec } = processService
+  return (scriptName: string, args: ReadonlyArray<string>) =>
+    gen(function* ($) {
+      const projectDir = yield* $(ensureProjectDir(processService, projectDirRef))
+      return yield* $(exec(`${command} ${scriptName} ${args.join(' ')}`.trim(), projectDir))
+    })
+}
+
 /**
  * Returns a function that executes a package manager command in the project directory.
  */
 export const makeScopedRun = (command: string, projectDirRef: Ref.Ref<string>) =>
   gen(function* ($) {
     const processService = yield* $(ProcessService)
-    const { exec } = processService
-    return (scriptName: string, args: ReadonlyArray<string>) =>
-      gen(function* ($) {
-        const projectDir = yield* $(ensureProjectDir(processService, projectDirRef))
-        return yield* $(exec(`${command} ${scriptName} ${args.join(' ')}`.trim(), projectDir))
-      })
+    return effectfulRun(processService, command, projectDirRef)
+  })
+
+/**
+ * Returns a function that executes the proper build command in the project directory.
+ */
+export const makeScopedBuild = (command: string, projectDirRef: Ref.Ref<string>) =>
+  gen(function* ($) {
+    const processService = yield* $(ProcessService)
+    const fileSystemService = yield* $(FileSystemService)
+    const scriptExists = yield* $(checkScriptExists(processService, fileSystemService, 'compile'))
+    if (scriptExists) {
+      return effectfulRun(processService, command, projectDirRef).bind(null, 'compile')
+    }
+    return effectfulRun(processService, command, projectDirRef).bind(null, 'build')
```

Comment:
Generally I try to avoid `if` statements in effectful code, not because of anything special, but because I believe everything in the same level of indentation makes it easier to read, as the syntax is already a bit noisy, why not:

```suggestion
    const script = scriptExists ? 'compile' : 'build'
    return effectfulRun(processService, command, projectDirRef).bind(null, script)
```

Also, effect tends to favor curried functions. In fact, most of it's API is like that, I'd say that its better to curry functions rather than using `bind`

Reply:
I tried to remove the `bind`, but that made the use of this function less clear in other places and tests, so better leaving it there for now 😅